lib/Dialect/FIRRTL/Transforms/PrefixModules.cpp

```diff
@@ -0,0 +1,177 @@
+//===- PrefixModules.cpp - Prefix module names pass -------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the PrefixModules pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "./PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/InstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Support/LLVM.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/SmallSet.h"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+
+/// This is the prefix which will be applied to a module.
+struct PrefixInfo {
+
+  /// The string to prefix on to the module and all of its children.
+  StringRef prefix;
+
+  /// If true, this prefix applies to the module itself.  If false, the prefix
+  /// only applies to it's children.
+  bool inclusive;
+};
+
+/// This maps a FModuleOp to a list of all prefixes that need to be applied.
+/// When a module has multiple prefixes, it will be cloned for each one.
+using PrefixMap = DenseMap<StringRef, llvm::SmallSet<std::string, 1>>;
```

Comment:
I'm not following all the logic here, but this is an extremely expensive representation for this.  Do you really need a SmallSet here instead of an array like thing?  If duplicates are possible, you could toss all the things into the array then sort/eliminate dupes before using it.

Also, do you need string here, or can you use a StringRef into the underlying attributes?

You might also consider StringMap<x> instead of DenseMap<StringRef, x>

Reply:
A Set is logically what I need here, but I could try to using a vector.  The way this works, it will try to insert a new prefix in to the list for each instance of a module. In the majority of cases, each module is getting renamed with a single prefix (i.e. set size of 1) so maybe I can do a linear scan on insertion.